src/gsi/read_prepbufr.f90

```diff
@@ -2963,51 +2998,76 @@ subroutine read_prepbufr(nread,ndata,nodata,infile,obstype,lunout,twindin,sis,&
         deallocate(presl_thin)
         call del3grids
      endif
-    if (.not.use_all_tm) then
-       deallocate(presl_thin)
+     if(.not.use_all_tm) then
+        deallocate(presl_thin)
         call del3grids_tm
      endif
 
-
 ! Normal exit
 
   enddo loop_convinfo! loops over convinfo entry matches
   deallocate(lmsg,tab,nrep)
+
 ! Close unit to bufr file
   call closbf(lunin)
   close(lunin)
 
-! Apply hilbert curve for cross validation if requested
-
-    if(lhilbert) &
-       call apply_hilbertcurve(maxobs,obstype,cdata_all(thisobtype_usage,1:maxobs))   
-
-! Write header record and data to output file for further processing
-  allocate(iloc(ndata))
-  icount=0
-  do i=1,maxobs
-     if(isort(i) > 0)then
-       icount=icount+1
-       iloc(icount)=isort(i)
+  nxdata=ndata
+  ndata=0
+  if(nxdata > 0)then
+!    numthin=0
+!    numqc=0
+!    numrem=0
+!    do i=1,maxobs
+!       if(.not. rusage(i))then
+!          numqc=numqc+1
+!       else if(rthin(i))then
+!          numthin=numthin+1
+!       else
+!          numrem=numrem+1
+!       end if
+!    end do
+!    write(6,*) ' prep ',trim(ioctype(nc)),ictype(nc),icsubtype(nc),numall,numrem,numqc,numthin
+!   If thinned data set quality mark to 14
+        !  If flag to not save thinned data is set - compress data
+     do i=1,nxdata
+        !   pmot=0 - all obs - thin obs
+        !   pmot=1 - all obs
+        !   pmot=2 - use obs
+        !   pmot=3 - use obs + thin obs
+                
+        if((pmot == 0 .and. .not. rthin(i)) .or. &
+           (pmot == 1) .or. &
+           (pmot == 2 .and. (rusage(i) .and. .not. rthin(i)))  .or. &
+           (pmot == 3 .and. rusage(i))) then
+    
+           if(rthin(i) .and. iqm > 0)cdata_all(iqm,i)=14
+           if(.not. rusage(i))cdata_all(iuse,i) = max(cdata_all(iuse,i),101.0_r_kind)
+           ndata=ndata+1
+           if(i > ndata)then
+              do k=1,nreal
+                 cdata_all(k,ndata)=cdata_all(k,i)
+              end do
+           end if
+        end if
+     end do
+     if(uvob)then
+        nodata=nodata+2*ndata
+     else
+        nodata=nodata+ndata
      end if
-  end do
-  if(ndata /= icount)then
-     write(6,*) ' PREPBUFR: mix up in read_prepbufr ,ndata,icount ',ndata,icount
-     call stop2(50)
   end if
-  allocate(cdata_out(nreal,ndata))
-  do i=1,ndata
-     itx=iloc(i)
-     do k=1,nreal
-        cdata_out(k,i)=cdata_all(k,itx)
-     end do
-  end do
-  deallocate(iloc,isort,cdata_all)
+  deallocate(rusage,rthin)
+
+! Apply hilbert curve for cross validation if requested
+
+  if(lhilbert) &
+       call apply_hilbertcurve(maxobs,obstype,cdata_all(thisobtype_usage,1:ndata))   
```

Comment:
At this point, the data should be compressed from 1:maxobs to 1:ndata.  So I think this is OK (at least as correct as the prior code).  However, maxobs in the call should be changed to ndata.

I also thought that the location of the call to apply_hilbertcurve may be in the wrong location.  However, if it is the prior code  is also wrong.  After looking at the routine, I believe the code actually locates the corresponding information from the accum_hilbertcurve.  So I think everything is OK.  It would be good if this were checked again.